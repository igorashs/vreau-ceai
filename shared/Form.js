import styled from 'styled-components';

export const Form = styled.form`
  display: flex;
  flex-direction: column;
  gap: calc(var(--baseline) / 2);
`;

export const FormAction = styled.div`
  margin-top: calc(var(--baseline) / 2);
  display: flex;

  ${({ justify }) => {
    switch (justify) {
      case 'space-between':
        return 'justify-content: space-between;';
      case 'flex-end':
        return 'justify-content: flex-end;';
      case 'flex-start':
        return 'justify-content: flex-start;';
      case 'center':
        return 'justify-content: center;';
      default:
        return 'justify-content: stretch;';
    }
  }}

  ${({ align }) => {
    switch (align) {
      case 'flex-end':
        return 'align-items: flex-end;';
      case 'flex-start':
        return 'align-items: flex-start;';
      case 'center':
        return 'align-items: center;';
      default:
        return 'align-items: stretch;';
    }
  }}
`;
