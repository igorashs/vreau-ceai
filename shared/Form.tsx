import styled from 'styled-components';

export const Form = styled.form`
  display: flex;
  flex-direction: column;
  gap: calc(var(--baseline) / 2);
`;

interface FormActionProps {
  justify?: 'space-between' | 'flex-end' | 'flex-start' | 'center';
  align?: 'flex-end' | 'flex-start' | 'center';
}

export const FormAction = styled.div<FormActionProps>`
  margin-top: calc(var(--baseline) / 2);
  display: flex;

  ${({ justify }) => {
    switch (justify) {
      case 'space-between':
        return 'justify-content: space-between;';
      case 'flex-end':
        return 'justify-content: flex-end;';
      case 'flex-start':
        return 'justify-content: flex-start;';
      case 'center':
        return 'justify-content: center;';
      default:
        return 'justify-content: stretch;';
    }
  }}

  ${({ align }) => {
    switch (align) {
      case 'flex-end':
        return 'align-items: flex-end;';
      case 'flex-start':
        return 'align-items: flex-start;';
      case 'center':
        return 'align-items: center;';
      default:
        return 'align-items: stretch;';
    }
  }}
`;
