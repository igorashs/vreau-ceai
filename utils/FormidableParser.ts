import * as validator from '@/utils/validator';
import FormidableNameSpace, { Files, File, IncomingForm } from 'formidable';
import { NextApiRequest } from 'next';
import { Formidable, ProductFields } from 'types';
import { productMessages } from './validator/schemas/product';
import { NextApiSessionRequest } from './withSession';

type ParsedData = {
  fields: Partial<ProductFields>;
  files: Files & { src?: File };
};

type HandleProductForm = (
  req: NextApiRequest | NextApiSessionRequest,
  options: Partial<FormidableNameSpace.Options>,
  allowedFileTypes: string[],
) => Promise<ParsedData>;

const handleProductForm: HandleProductForm = async (
  req,
  options,
  allowedFileTypes,
) => {
  const data: ParsedData = await new Promise((resolve, reject) => {
    const form: Formidable = new IncomingForm(options);

    form.onPart = (part) => {
      if (part.filename) {
        // get file extension
        const fileType = part.mime?.split('/').pop() ?? '';

        // validate allowed file types
        if (!allowedFileTypes.includes(fileType)) {
          form.emit(
            'error',
            validator.createValidationError({
              message: productMessages.src.invalid,
              key: 'src',
            }),
          );
        }
      }

      form.handlePart(part);
    };

    form.parse(req, (err, fields, files) => {
      if (err) {
        // custom error on maxFileSize
        if (err.message.includes('maxFileSize')) {
          reject(
            validator.createValidationError({
              message: productMessages.src.max,
              key: 'src',
            }),
          );
        } else {
          reject(err);
        }
      }

      resolve({ fields, files });
    });
  });

  return data;
};

export default {
  handleProductForm,
};
