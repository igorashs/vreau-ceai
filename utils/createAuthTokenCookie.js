import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createAuthTokenCookie = (payload) => {
  const authToken = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  return cookie.serialize('auth-token', authToken, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });
};
