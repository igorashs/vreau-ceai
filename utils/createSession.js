import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createSession = (payload) => {
  const token = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  const [header, jwtPayload, signature] = token.split('.');
  const session = header + '.' + jwtPayload;

  const sessionCookie = cookie.serialize('session', session, {
    secure: process.env.NODE_ENV !== 'development',
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  const authCookie = cookie.serialize('auth', signature, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  return [sessionCookie, authCookie];
};
