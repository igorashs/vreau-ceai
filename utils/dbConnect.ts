import mongoose from 'mongoose';

const dbConnect = async () => {
  if (mongoose.connection.readyState >= 1) {
    return;
  }

  await mongoose.connect(process.env.MONGODB_URI as string, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
    useFindAndModify: false,
  });
};

export default dbConnect;
