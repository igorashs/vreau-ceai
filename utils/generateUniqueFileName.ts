import { nanoid } from 'nanoid';

const generateUniqueFileName = (fileName: string) => {
  // grab file extension
  const matches = fileName?.match(/\..+$/);

  // create new file name
  const uniqueFileName = nanoid() + (matches && matches[0]);

  return uniqueFileName;
};

export default generateUniqueFileName;
