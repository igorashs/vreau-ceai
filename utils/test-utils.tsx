import { render, RenderOptions } from '@testing-library/react';
import GlobalStyle from 'GlobalStyle';

const Wrapper: React.FC = ({ children }) => (
  <>
    <GlobalStyle />
    {children}
  </>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';

export { customRender as render };
