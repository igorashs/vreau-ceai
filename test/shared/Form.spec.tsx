import { render } from '@/utils/test-utils';
import Form, { FormAction } from '@/shared/Form';

describe('Form', () => {
  it('renders correctly', () => {
    const { container, getByText } = render(<Form>content</Form>);

    expect(getByText('content')).toBeVisible();
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('FormAction', () => {
  it('renders correctly', () => {
    const { container, getByText } = render(<FormAction>content</FormAction>);

    const formAction = getByText('content');

    expect(formAction).toBeVisible();
    expect(formAction).toHaveStyleRule('justify-content', 'stretch');
    expect(formAction).toHaveStyleRule('align-items', 'stretch');
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('justify prop', () => {
    it('renders with space-between', () => {
      const { container } = render(
        <FormAction justify="space-between">content</FormAction>,
      );

      expect(container.firstChild).toHaveStyleRule(
        'justify-content',
        'space-between',
      );
    });

    it('renders with flex-end', () => {
      const { container } = render(
        <FormAction justify="flex-end">content</FormAction>,
      );

      expect(container.firstChild).toHaveStyleRule(
        'justify-content',
        'flex-end',
      );
    });

    it('renders with flex-start', () => {
      const { container } = render(
        <FormAction justify="flex-start">content</FormAction>,
      );

      expect(container.firstChild).toHaveStyleRule(
        'justify-content',
        'flex-start',
      );
    });

    it('renders with center', () => {
      const { container } = render(
        <FormAction justify="center">content</FormAction>,
      );

      expect(container.firstChild).toHaveStyleRule('justify-content', 'center');
    });
  });

  describe('align prop', () => {
    it('renders with flex-end', () => {
      const { container } = render(
        <FormAction align="flex-end">content</FormAction>,
      );

      expect(container.firstChild).toHaveStyleRule('align-items', 'flex-end');
    });

    it('renders with flex-start', () => {
      const { container } = render(
        <FormAction align="flex-start">content</FormAction>,
      );

      expect(container.firstChild).toHaveStyleRule('align-items', 'flex-start');
    });

    it('renders with center', () => {
      const { container } = render(
        <FormAction align="center">content</FormAction>,
      );

      expect(container.firstChild).toHaveStyleRule('align-items', 'center');
    });
  });
});
