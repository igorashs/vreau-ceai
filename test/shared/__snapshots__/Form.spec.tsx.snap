// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: calc(var(--baseline) / 2);
}

<form
  class="c0"
>
  content
</form>
`;

exports[`FormAction renders correctly 1`] = `
.c0 {
  margin-top: calc(var(--baseline) / 2);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: stretch;
  -webkit-justify-content: stretch;
  -ms-flex-pack: stretch;
  justify-content: stretch;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

<div
  class="c0"
>
  content
</div>
`;
